File: sorting/selection_sort.cc
/**
 * @file selection_sort.cc
 * @author Hongyu Ding
 * @brief This is the code of the selection sort
 * @version 1.0
 * @date 2020-10-17
 * 
 * @copyright Copyright (c) 2020
 * 
 */
#include <iostream>
#include <vector>

/**
 * @brief selection_sort sorts the array a
 * O(n^2)
 * 
 * @param a the array we want to perform selection sort with
 */
void selection_sort(std::vector<int> &a) {
    int len = a.size();
    for (int i = 0; i < len - 1; i ++) {
        int min = i;
        for (int j = i + 1; j < len; j ++) {
            if (a[j] < a[min]) {
                min = j;
            }
        }
        std::swap(a[i], a[min]);
    }
}

int main() {
    std::vector<int> array;
    int n;
    
    // input
    std::cin >> n;
    for (int i = 0; i < n; i++) {
        int val;
        std::cin >> val;
        array.push_back(val);
    }

    selection_sort(array);

    // output
    for (int i = 0; i < n; i++) {
        std::cout << array[i] << " ";
    }
    std::cout << std::endl;

    return 0;
}

File: sorting & searching/Sort&SearchJavaVersion/SelectionSort.java
/**
 * @file SelectionSort.java
 * @author Weilan Tao
 * @brief This is the code of the java version Selection Sort
 * @version 1.0
 * @date 2020-10-18 EST
 *
 * @copyright Copyright (c) 2020
 *
 */


import java.util.*;
public class SelectionSort {

    /**
     * Selection Sort
     * @param array
     * @param len
     *O(n^2)
     */
    public static void selectionSort(int[] array, int len){
        for(int i=0;  i<len ; i++){//loop through the array 0~length-1
            int minIndex=i;

            for (int j=i; j<len;j++){//loop through the unsorted part i~length-1
                if(array[j]<array[i]){
                    array[i]=array[j];
                    minIndex=j;
                }
            }

            int temp=array[i];
            array[i]=array[i];
            array[minIndex]=temp;

        }

    }


    public static void main(String args[]){
        System.out.println("enter the length of an array");
        Scanner s1= new Scanner(System.in);
        int x;
        x=s1.nextInt();

        int[] array= new int[x];
        System.out.println("set the array");

        for(int i=0; i<x; i++){
            Scanner s2= new Scanner(System.in);
            array[i]=s2.nextInt();
        }

        selectionSort(array,x);

        System.out.println("print the sorted array");
        for(int i=0; i<x; i++){
            System.out.print(array[i]);
        }
    }

}
